staging/src/k8s.io/client-go/util/workqueue/delaying_queue.go

```diff
@@ -52,13 +101,13 @@ type DelayingQueueConfig struct {
 // NewDelayingQueue constructs a new workqueue with delayed queuing ability.
 // NewDelayingQueue does not emit metrics. For use with a MetricsProvider, please use
 // NewDelayingQueueWithConfig instead and specify a name.
-func NewDelayingQueue() DelayingInterface {
+func NewDelayingQueue() *delayingType {
```

Comment:
I think it is not a good practice to return a private type from a public function.

Reply:
Can you provide any links to documentation about this?  I did have a look and could not find this.  I thought it was quite common to return a private type when you want to ensure that people use your constructor and not make their own.  In this case, we have exported the DelayingInterface (and ExtendedDelayingInterface) which is what they will likely store their version of the delayingType into.

Reply:
I mean we would be returning a type which is not exported and does not give a clear contract to the user what to expect. So using an interface seems like a better option.

Even if GO would allow you to use it outside of its intended package, you cannot create a following function
`func test(a workqueue.delayingType) `  where you would pass the queue. This will result with a linter complaining `Usage of the unexported type 'delayingType'`

I do not know if they have it documented somewhere, but IMO it is a not good practice.

Reply:
But we can still use it in the tests which shares the same package

Reply:
> Even if GO would allow you to use it outside of its intended package, you cannot create a following function
func test(a workqueue.delayingType)  where you would pass the queue. This will result with a linter complaining Usage of the unexported type 'delayingType'

The point here is that people are unable to use the type outside of this package (because then they will have a dependency on our implementation).  They are forced to use our constructor which is a good thing.  They can store the queue in a DelayingInterface variable or struct as before and then only the methods of the interface are available to them.  They would, for example, not be able to call the syncFill() method.